how-to/customize-workspace/client/src/connections.ts

```diff
@@ -0,0 +1,229 @@
+						`This connection has specified payload validation but that check needs to be implemented. UUID: ${validatedConnection.identity.uuid}`
+					);
+					isValid = true;
+				}
+				if (isValid) {
+					if (connectedClients[identity.uuid] !== undefined) {
+						isValid = false;
+						errorMessage =
+							"This platform can only accept one connection from an external connection. The uuid is already registered.";
+					} else {
+						// assign the passed identity
+						validatedConnection.identity = identity;
+						connectedClients[identity.uuid] = validatedConnection;
+						console.log(`The following connection has been added to the connected list: ${identity.uuid}`);
+					}
+				}
+			}
+
+			if (!isValid) {
+				console.warn(
+					`The following connection has not been added to the connected list: ${identity.uuid} as it failed validation`
+				);
+				throw new Error(errorMessage);
+			}
+		});
+
+		connectionService.onDisconnection(async (identity) => {
+			console.log(`Client disconnected uuid: ${identity.uuid}, name: ${identity.name}`);
+			await disconnect(identity);
+		});
+
+		connectionService.register("action", async (payload, identity) => {
+			console.log("Action received from client:", identity, payload);
+			const result = await executeAction(identity, payload as { action: string });
+			return { result };
+		});
+
+		connectionService.register("canAction", async (payload, identity) => {
+			console.log("Check for action permission received from client:", identity, payload);
+			const result = isActionSupported(identity, payload as { action: string });
+			return { result };
+		});
+	} else {
+		console.log(`This platform is not configured to support a connectionProvider.`);
+	}
+}
+
+export function registerAction(actionName: string, action: () => Promise<void>): boolean {
+	if (registeredActions[actionName] === undefined) {
+		console.log(`Adding action ${actionName} to available actions list.`);
+		registerAction[actionName] = action;
+		return true;
+	}
+	console.warn(`Not adding action ${actionName} to available actions list as it is already registered.`);
+	return false;
+}
+
+export function clearAction(actionName: string): boolean {
+	if (registeredActions[actionName] === undefined) {
+		console.warn(`Cannot remove action ${actionName} from available actions list as it is not registered.`);
+		return false;
+	}
+	delete registeredActions[actionName];
+	console.log(`Action ${actionName} cleared from available actions list.`);
+	return true;
+}
+
+export async function getConnectedAppSourceClients() {
+	const connections: ConnectedAppSourceConnection[] = [];
+
+	const availableConnections = Object.values(connectedClients);
+	for (const connection of availableConnections) {
+		const matchedConnection = connection.connectionTypes.find(
+			(supportedConnectionType) => supportedConnectionType.type === "appSource"
+		);
+
+		if (matchedConnection !== undefined && matchedConnection.type === "appSource") {
```

Comment:
change to optional chain `matchedConnection?.type`

Reply:
updated for both appSource and other sources.